File: JavaScript/Easy/BuddyStrings.js
/**
 * 859. Buddy Strings
 * @param {string} s
 * @param {string} goal
 * @return {boolean}
 */
var buddyStrings = function (s, goal) {
  if (s.length != goal.length) return false;

  if (s === goal) {
    let count = new Array(26).fill(0);
    for (let i = 0; i < goal.length; i++) {
      count[goal.charCodeAt(i) - "a".charCodeAt(0)]++;
      if (count[goal.charCodeAt(i) - "a".charCodeAt(0)] > 1) {
        return true;
      }
    }
    return false;
  } else {
    let first = -1, second = -1;
    for (let i = 0; i < s.length; ++i) {
      if (s[i] !== goal[i]) {
        if (first === -1) first = i;
        else if (second === -1) second = i;
        else return false;
      }
    }
    return second !== -1 && s[first] === goal[second] && s[second] === goal[first];
  }
};

File: Java/src/com/leetcode/Easy/BuddyStrings.java
package com.leetcode.Easy;

// 859. Buddy Strings
public class BuddyStrings {
    public boolean buddyStrings(String A, String B) {
        if (A.length() != B.length()) return false;

        if (A.equals(B)) {
            int[] count = new int[26];
            for (int i = 0; i < B.length(); i++) {
                count[B.charAt(i) - 'a']++;
            }
            for (int j : count) {
                if (j > 1)
                    return true;
            }
            return false;
        } else {
            int first = -1, second = -1;
            for (int i = 0; i < A.length(); ++i) {
                if (A.charAt(i) != B.charAt(i)) {
                    if (first == -1)
                        first = i;
                    else if (second == -1)
                        second = i;
                    else
                        return false;
                }
            }
            return (second != -1 && A.charAt(first) == B.charAt(second) && A.charAt(second) == B.charAt(first));
        }
    }
}
